src/storage/src/render/upsert.rs

```diff
@@ -285,25 +370,40 @@ where
                             .expect("key missing from commands_state");
                         match value {
                             Some(value) => {
-                                if let Some(old_value) = command_state.replace(value.clone()) {
+                                if let Some(old_value) = command_state.update_value(value.clone()) {
                                     output_updates.push((old_value, ts.clone(), -1));
                                 }
                                 output_updates.push((value, ts, 1));
                             }
                             None => {
-                                if let Some(old_value) = command_state.take() {
+                                if let Some(old_value) = command_state.remove_value() {
                                     output_updates.push((old_value, ts, -1));
                                 }
                             }
                         }
                     }
 
+                    // Accumulating metrics before draining
+                    let total_diff_records = commands_state
+                        .values()
+                        .map(|v| Into::<i64>::into(v.diff_record))
+                        .sum();
+                    let total_diff_bytes = commands_state.values().map(|v| v.diff_bytes).sum();
+
                     // Record the changes in `state`.
                     state
-                        .multi_put(commands_state.drain(..))
+                        .multi_put(commands_state.drain(..).map(|(k, v)| (k, v.value)))
```

Comment:
You can calculate this without an extra iteration i think:
```suggestion
                        let mut total_diff_records = 0;
                         let mut total_diff_bytes = 0;
                        state.multi_put(commands_state.drain(..).map(|(k, v)| {
                        	   total_diff_records += v.diff_record;
                        	   total_diff_bytes += v.diff_bytes;
                            (k, v.value)
           				})
                         )
```